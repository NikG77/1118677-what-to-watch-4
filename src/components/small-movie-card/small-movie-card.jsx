import React from "react";
import PropTypes from "prop-types";

const SmallMovieCard = (props) => {
  const {film, onSmallMovieCardClick, onSmallMovieCardHover} = props;
  const {title, src: srcFilm} = film;

  return (
    <article
      onMouseEnter={() => {
        onSmallMovieCardHover(film);
      }} className="small-movie-card catalog__movies-card">
      <div className="small-movie-card__image">
        <img src={srcFilm} alt={title} width="280" height="175" />
      </div>
      <h3 onClick={onSmallMovieCardClick}

        className="small-movie-card__title">
        <a className="small-movie-card__link" href="movie-page.html">{title}</a>
      </h3>
    </article>

  );
};

SmallMovieCard.propTypes = {
  film: PropTypes.shape({
    title: PropTypes.string.isRequired,
    src: PropTypes.string.isRequired,
  }).isRequired,
  onSmallMovieCardClick: PropTypes.func.isRequired,
  onSmallMovieCardHover: PropTypes.func.isRequired,
};

export default SmallMovieCard;

