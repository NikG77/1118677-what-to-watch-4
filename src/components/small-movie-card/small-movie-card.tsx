import * as React from "react";
import {connect} from "react-redux";
import {Link} from "react-router-dom";

import {ActionCreator} from "../../reducer/watch/watch";
import {AppRoute, TIME_DELAY} from "../../const";
import {FilmType} from "../../types";
import history from "../../history";

import VideoPlayer from "../video-player/video-player";


let timerId;

const setTimer = (film, cb) => {
  timerId = setTimeout(() => cb(film), TIME_DELAY);
};

const clearTimer = (film, cb) => {
  clearTimeout(timerId);
  cb(film);
};

interface Props {
  film: FilmType;
  onSmallMovieCardClick: () => void;
  onSmallMovieCardHover: () => void;
  isPlaying: boolean;
}

const SmallMovieCard: React.FunctionComponent<Props> = (props: Props) => {
  const {
    film,
    onSmallMovieCardClick,
    onSmallMovieCardHover,
    isPlaying
  } = props;

  const {
    id,
    previewVideo,
    title,
    src
  } = film;

  return (
    <article className="small-movie-card catalog__movies-card"
      onMouseEnter={() => setTimer(film, onSmallMovieCardHover)}
      onMouseLeave={() => clearTimer(null, onSmallMovieCardHover)}
      onClick={() => {
        clearTimer(film, onSmallMovieCardClick);
        history.push(`${AppRoute.FILM}/${id}`);
      }}
    >
      <div className="small-movie-card__image">
        <VideoPlayer
          src={previewVideo}
          poster={src}
          isPlaying={isPlaying}
        />
      </div>
      <h3 className="small-movie-card__title">
        <Link to={`${AppRoute.FILM}/${id}`} className="small-movie-card__link">{title}</Link>
      </h3>
    </article>
  );
};


const mapDispatchToProps = (dispatch) => ({
  onSmallMovieCardClick(film) {
    dispatch(ActionCreator.setFilm(film));
    dispatch(ActionCreator.setGenre(film.genre));
  }
});

export {SmallMovieCard};

export default connect(null, mapDispatchToProps)(SmallMovieCard);
