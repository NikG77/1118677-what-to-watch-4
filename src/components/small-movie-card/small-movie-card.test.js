import React from "react";
import renderer from "react-test-renderer";
import {SmallMovieCard} from "./small-movie-card";
import history from "../../history";
import {Router} from "react-router-dom";

const film = {
  title: `One Flew Over the Cuckoo's Nest`,
  src: ``,
  poster: ``,
  ratingScore: 8,
  ratingCount: 200,
  director: ``,
  starring: [``, ``],
  genre: `Drama`,
  releaseDate: 2020,
  pictureBackground: ``,
  previewVideo: ``,
  duration: 100,
  backgroundColor: ``,
  videoLink: ``,
  description: ` `,
  isFavorite: false,
  id: 1,
};

it(`Should SmallMovieCard render correctly`, () => {
  const tree = renderer
    .create(
        <Router
          history={history}
        >
          <SmallMovieCard
            film={film}
            onSmallMovieCardClick={() => {}}
            onSmallMovieCardHover={() => {}}
            isPlaying={false}
          />
        </Router>,
        {
          createNodeMock: () => {
            return {};
          }
        }).toJSON();

  expect(tree).toMatchSnapshot();
});
