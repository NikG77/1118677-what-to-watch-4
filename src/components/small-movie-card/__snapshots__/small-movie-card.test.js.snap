// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should SmallMovieCard render correctly 1`] = `
<article
  className="small-movie-card catalog__movies-card"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="small-movie-card__image"
  >
    <video
      height="175"
      width="280"
    />
  </div>
  <h3
    className="small-movie-card__title"
  >
    <a
      className="small-movie-card__link"
      href="/films/1"
      onClick={[Function]}
    >
      One Flew Over the Cuckoo's Nest
    </a>
  </h3>
</article>
`;
